File: guess_api/lib/domain/chess_piece_custom_moviment.dart
import 'package:guess_api/domain/chess_piece_moviment.dart';
import 'package:guess_api/models/chess_packet.dart';
import 'package:guess_api/models/chess_piece.dart';

class ChessPieceCustomMoviment{
  final ChessPieceMoviment _data;
  ChessPieceCustomMoviment.upgrade(this._data){
    if(!_data.isCustom){
      throw "This moviment isn't a custom moviment!";
    }
  }

  List<ChessPacket>? execMoviment(int player, ChessPiece piece,
    ChessPiece? Function(int, int) pieceRequester, bool Function(int, int, int, int) hasPieceInWay){
    switch(this._data){
      case ChessPieceMoviment.rockRight:
        ChessPiece? tower = pieceRequester(piece.x + (player == 1 ? 4 : 3), piece.y);
        if(tower == null || tower.destroyed){
          return null;
        }

        if(tower.moved || hasPieceInWay(piece.x, piece.y, tower.x, tower.y)){
          return null;
        }

        piece.updatePosition(piece.x + 2, piece.y);
        tower.updatePosition(piece.x - 1, tower.y);

        return [ChessPacket.updateChessPiecePosition(piece.id, piece.x, piece.y),
          ChessPacket.updateChessPiecePosition(tower.id, tower.x, tower.y)];
      case ChessPieceMoviment.rockLeft:
        ChessPiece? tower = pieceRequester(piece.x - (player == 1 ? 3 : 4), piece.y);
        if(tower == null || tower.destroyed){
          return null;
        }

        if(tower.moved || hasPieceInWay(piece.x, piece.y, tower.x, tower.y)){
          return null;
        }

        piece.updatePosition(piece.x - 2, piece.y);
        tower.updatePosition(piece.x + 1, tower.y);

        return [ChessPacket.updateChessPiecePosition(piece.id, piece.x, piece.y),
          ChessPacket.updateChessPiecePosition(tower.id, tower.x, tower.y)];
      default:
        return null;
    }
  }
}

File: guess/js/domain/chess_piece_custom_moviment.js
class ChessPieceCustomMoviment{
    static getCustomMovimentationMark(player, piece_position, movimentName){
        if(movimentName == "rock_right"){
            return {
                "x": (player == 1 ? 4 : 3),
                "y": 0
            }
        }
        if(movimentName == "rock_right"){
            return {
                "x": (player == 1 ? -3 : -4),
                "y": 0
            }
        }
        return {
            "x": 0,
            "y": 0
        }
    }

    static canDoThisMoviment(player, piece, moviment, pieceRequester, hasPieceInWay){
        if(moviment["name"] == "rock_right"){
            let piecePosition = piece["piece_position"]
            let tower = pieceRequester(piecePosition["x"] + (player == 1 ? 4 : 3), piecePosition["y"])

            if(!tower || tower["piece_type"] != "tower"){
                return false
            }

            if(hasPieceInWay(piecePosition["x"], piecePosition["y"], tower["piece_position"]["x"], tower["piece_position"]["y"])){
                return false
            }

            return true
        }
        if(moviment["name"] == "rock_left"){
            let piecePosition = piece["piece_position"]
            let tower = pieceRequester(piecePosition["x"] - (player == 1 ? 3 : 4), piecePosition["y"])
            
            if(!tower || tower["piece_type"] != "tower"){
                return false
            }

            if(hasPieceInWay(piecePosition["x"], piecePosition["y"], tower["piece_position"]["x"], tower["piece_position"]["y"])){
                return false
            }

            return true
        }
        return false;
    }
}